### PR TITLE
[AOTI] Support module buffer mutation

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122824

Summary: Fixes https://github.com/pytorch/pytorch/issues/120424. Because in a forward pass module buffers may be mutated, we need to allow that in AOTI. In addition, this will be a necessary step if we want to extend AOTI to training.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @chauhang

Differential Revision: [D55528912](https://our.internmc.facebook.com/intern/diff/D55528912)

### CODE REVIEW

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55528912).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55528912).

**Comment by desertfire** [comment]
Discussed with @bdhirsh offline. This PR doesn't solve the problem where there is an alias between buffers and a mutation happens on one of them. With the current AOTI usage scenario, that case will fail to export in the first place.

To properly support alias, we need to lift weights into args, leverage existing functionalization, and separate model code and data in the generated .so file. Discussed with @muchulee8, and he is going to look into that.

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -1679,14 +1679,6 @@ def _compile_consts_linux(consts: bytes) -> str:
...
             log.debug("aot constant binary command: %s", cmd)
 
-            cmd = (
-                f"{objcopy_command} --rename-section"
-                " .data=.lrodata,alloc,load,readonly,data,contents"
-                f" {consts_o} {consts_o}"
-            )
-            log.debug("aot constant obj command: %s", cmd)
-            run_command_and_check(cmd)
-
```
Comment (chenyang78) [inline torch/_inductor/codecache.py:1689]: Wondering why we are removing this command?

File: torch/_inductor/compile_fx.py [hunk 2]
```diff
@@ -160,12 +164,23 @@ def _unlift_graph(mod, gm, graph_signature):
...
-            lifted_inputs.append(graph_signature.inputs_to_parameters[node_name])
+            parameter_name = graph_signature.inputs_to_parameters[node_name]
+            lifted_inputs.append(parameter_name)
         elif node_name in graph_signature.inputs_to_buffers:
-            lifted_inputs.append(graph_signature.inputs_to_buffers[node_name])
+            buffer_name = graph_signature.inputs_to_buffers[node_name]
+            lifted_inputs.append(buffer_name)
+            gm.meta[
+                get_cloned_parameter_buffer_name(buffer_name)
+            ] = clone_preserve_strides(state_dict[buffer_name])
```
Comment (chenyang78) [inline torch/_inductor/compile_fx.py:183]: If two buffers are aliased (would this be possible?) , cloning would break such aliasing?

**Review by jansel — APPROVED** [review]

File: torch/_inductor/compile_fx.py (hunk 2 above)
Comment (desertfire) [reply torch/_inductor/compile_fx.py:183]: Yes, you are right. But as I commented in this PR main page, that case will fail to export in today's AOTI use case.  We are thinking to handle that indeed.

**Review by chenyang78 — APPROVED** [review]

**Comment by malfet** [comment]
I guess you'll need to modify to make weights mutable on Darwin:
https://github.com/pytorch/pytorch/blob/d765e223acec89e3795e6cab626b23c620578fc0/torch/_inductor/codecache.py#L1719

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55528912).

**Pull request closed without merging**